README
# $Id: README,v 1.3 1995/12/18 10:12:25 jhi Exp $
# Time-stamp:	<95/12/18 12:12:14 jhi>

This Perl extension implements the BSD process resource limit functions

	getrusage()	getrlimit()	setrlimit()

and the BSD process priority functions.

	getpriority()	setpriority()

NOTE 1: Perl 5 is required.

NOTE 2: if your operating system does not support these BSD functions,
Perl cannot help. In detailed terms: you _must_ have the C header file
<sys/resource.h>. If you do not, this package is of little use to you.

NOTE 3: The header file alone is not enough: you must have library
support also. The C compiler include paths (-I...) and library paths
(-L...) and the attempted libraries (-l...) are listed in Makefile.PL.

NOTE 4: More caveats (vendors sloppy support for these things, et. al.)
        in the Resource.pm inlined pod doc.

To compile

	perl Makefile.PL
	make

and to install

	make install

The documentation is the pod inlined into Resource.pm, use for example

	pod2man Resource.pm | nroff -man | more

to view the examples.

You can test how well the extension works with

	make test

Please report any bugs and/or missing/excess functionality to me.
Especially if your system does support these BSD functions but
the Makefile.PL fails to locate them, please let me know.

The platforms known to work are

	AIX 3.2.5
	AIX 4.1.1
	Digital UNIX 3.2C
	HP-UX 9.03
	IRIX 6.0.1
	SunOS 4.1.3
	SunOS 5.3

Jarkko Hietaniemi <Jarkko.Hietaniemi@hut.fi>

1995-12-18


ChangeLog
2013-07-12  Jarkko Hietaniemi  <jhi@Vredefort-2.local>

	* Address rt.cpan.org #86857: t/setrlimit.t failure
	  Skip setrlimit() tests also if the hard limit is RLIM_INFINITY.

	* Release 1.2907.

2013-07-07  Jarkko Hietaniemi  <jhi@iki.fi>

	* Skip number-of-files and stack setrlimit tests in Cygwin:
	  the constants exist but the implementation doesn't really
	  seem to be there.  Document this in the pod.  Testing by dsolimano.

	* For setrlimit tests involving byte sizes (memory or file size)
	  round the byte sizes down to 4 kB boundaries.

	* Fix some pod nits.
	
	* Release 1.2906.

2013-06-22  Jarkko Hietaniemi  <jhi@iki.fi>

	* Address rt.cpan.org #86334: typo fixes
	
	* Address rt.cpan.org #62504: Disable user/system
	  timing tests if workload accumulate too little time

	* Attempt harder in getrusage.t to burn enough user and system time,
	  refactor the "burn some cpu" utility into t/burn.pl, use that also
	  in t/times.t.
	
	* Fix a testing problem in NetBSD for t/setrlimit.t,
	  the tests were misnumbered.
	
	* Silence compilation warnings coming from Perl 5.18.0.
	
	* Clarify the license to be Artistic 2.0 or LGPL 2.0.
	
	* Regenerate ppport.h with Perl 5.18.0.
	
	* Release 1.2905.

2010-03-14  Jarkko Hietaniemi  <jhi@iki.fi>

	* Address rt.cpan.org #53856: memory leak in get_rlimits()

	* Address rt.cpan.org #53859: clarify get_rlimits() docs
	
	* Release 1.2904.

2009-03-31  Jarkko Hietaniemi  <jhi@iki.fi>

	* Address rt.cpan.org #44082:
	  "Bug in compiler detection (Makefile.PL)" (for AIX xlc)
	  Patch submitted in the report by rehsack at cpan.org.
	
	* Release 1.2903.

2008-12-30  Jarkko Hietaniemi  <jhi@iki.fi>

	* Lowered the "0.95 of the current limit" down to
	  "0.75 of the current limit" in setrlimit.t so that
	  more memory-constrained systems (like 2MB or less of
	  default stack) have slightly more chance of completing the test.
	  The 0.75 also has more chance of aligning with page sizes.
	
	* The scalar context return of getrlimit() was documented
	  to return an object, even though it really returned the
	  soft limit, reported by anicka@suse.cz.  Fixed the documentation.
	
	* Judging by the cpantesters reports no netbsd has a functional
	  RLIMIT_STACK, so skipping the test in all of them.
	
	* Release 1.2902.

2008-01-29  Jarkko Hietaniemi  <jhi@iki.fi>

	* netbsd-alpha does not have a functional RLIMIT_STACK,
	  reported by David Cantrell.  Skip the test, and document
	  as a known issue.
	
	* Release 1.2901.

2008-01-27  Jarkko Hietaniemi  <jhi@iki.fi>

	* Rewrite the setrlimit.t test.	 Simplify it a lot; now all
	  RLIM_INFINITY limits are simply skipped; for those that
	  are limited we try setting only the soft limit, and accept
	  a value lower than we tried to set, since various systems
	  might either enforce lower limits, or round up the limits.
	  Added documentation on setrlimit() about this.
	
	* Allow calling getrlimit/setrlimit and getpriority/setpriority
	  with the string names ("RLIMIT_NPROC", for example).

	* Include ppport.h (generated with perl 5.8.8), reshuffle
	  system header includes appropriately.

	* Tested also with older Perls, 5.005_05 and 5.6.2, in OS X.
	  (5.005_04 didn't work, it creates but then cannot dynaload
	   the .bundle files, go figure.)
	
	* Minor documentation tweaks and copyright year bumps.
	
	* Release 1.29.

2006-05-26  Jarkko Hietaniemi  <jhi@iki.fi>

	* Tweak the regexp parsing the ps -o output in setpriority.t,
	  from Alexey Tourbin and Rafael Garcia-Suarez.	 Still just
	  a best-effort attempt, but one can only try.

	* Release 1.28.

2006-05-25  Jarkko Hietaniemi  <jhi@iki.fi>

	* Address rt.cpan.org #13130 and #19412: try to cope with
	  an already reniced shell running the setpriority.t.  The
	  tricks used are trying "ps -o pid,nice" and looking for $$,
	  and if that fails, then trying whether the nice(1) is the
	  GNU one, and if so, running it without arguments.
	
	* Release 1.27.

2006-05-21  Jarkko Hietaniemi  <jhi@iki.fi>

	* Address rt.cpan.org #19298: bug in getrusage(): 'inblock' was
	  misspelt as 'inlock' (the end result being that people looking
	  for the 'inblock' field got zero)
	  
	* For Mac OS X reword the message about the RLIM_NPROC test failure
	  (or rather, the possible failure).  Still a mystery under what
	  exact conditions the test fails or succeeds.
	  
	* Add Test::Pod and Test::Pod::Coverage tests.
	
	* Release 1.26.

2006-04-09  Jarkko Hietaniemi  <jhi@iki.fi>

	* Address rt.cpan.org #13131: setrlimit.t tries to increase hard
	  limits, cannot raise limits higher than the hard limit.

	* Address rt.cpan.org #13130: getpriority.t and setpriority.t
	  assume priority = 0, solution hopefully portable.  Now should
	  work under e.g. "nice make test".
	
	* Add more verbosity to the case of the known rlimit.t test
	  failure in Mac OS X.
	
	* Release 1.25.

Sun Feb 29 16:51:34 2004  Jarkko Hietaniemi  <jhi@iki.fi>

	* No functional changes but document the known fact that
	  under at least Mac OS X 10.3.2 the t/setrlimit.t subtest #8
	  may fail because of an OS bug.  Also reorder the documentation
	  a bit to collect all the known issues under one heading.
	
	* Release 1.24.

Tue Oct	 7 21:54:04 2003  Jarkko Hietaniemi  <jhi@iki.fi>

	* Makefile.PL: use archlibexp instead of archlib in cc -I.../CORE
	  so that Perls installed under ~user work.

	* Release 1.23.

Sun Feb	 9 12:12:42 2003  Jarkko Hietaniemi  <jhi@iki.fi>

	* Add RLIMIT_LOCKS.

	* Fix a typo s/RLIM_SAVEWD_MAX/RLIMIT_SAVED_MAX/
	
	* Sort the @EXPORT list.
	
	* Release 1.22.

Thu Dec 12 01:15:44 2002  Jarkko Hietaniemi  <jhi@iki.fi>

	* Patches for NCR MP-RAS from grommel@sears.com.
	
	* Release 1.21.

Tue Nov 26 04:20:38 2002  Jarkko Hietaniemi  <jhi@iki.fi>


Makefile.PL
#
# Makefile.PL
#
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
#

require 5.002;

use vars qw($PACKAGE $DEFINE $LIBS);

$PACKAGE = 'BSD::Resource';

use Config;
use ExtUtils::MakeMaker;
use File::Basename;

$| = 1;

my $CC;

sub try_compile_and_link {
    my ($c, %opt) = @_;

    my ($ok) = 0;
    my ($tmp) = "tmp$$";
    local(*TMPC);

    my $cccmd;

    unlink("$tmp.c", "$tmp.o");

    if (open(TMPC, ">$tmp.c")) {
	print TMPC $c;
	close(TMPC);
	my @cc;
	if ($CC) {
	    push @cc, $CC;
	} else {
	    if (exists $ENV{CC}) {
		push @cc, $ENV{CC};
		print STDERR "(trying first the CC from your environment)\n";
	    }
	    push @cc, $Config{'cc'};
	}
	while (my $cc = shift @cc) {
	    unless (-x $cc) {
		# They might have a different compiler
		# than what Perl was compiled with (easy
		# if using gcc), or their $cc has more
		# than just the compiler name (say, options)
		my ($b, $s, $r) = ($cc =~ /^(\S+)(\s*)(.*)/);
		$cc = basename($b) . $s . $r;
	    }
	    $cccmd =
		"$cc -c -o $tmp.o -I$Config{'archlibexp'}/CORE $Config{'ccflags'}";
	    $cccmd .= " 2>/dev/null" unless $ENV{VERBOSE};
	    print STDERR "[cccmd = $cccmd]\n" if $ENV{VERBOSE};
	    if (system("$cccmd $tmp.c") == 0) {
		$ok = -s "$tmp.o";
		$CC = $cc;
	    }
	    last if $ok;
	}
	unlink("$tmp.c", "$tmp.o");
    }
    
    return $ok;
}

sub has_resource_h () {
    print "Checking whether you have certain header files...\n";

    if (try_compile_and_link(<<EOM)) {
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#include <sys/resource.h>
static int foo;
EOM
    print "You do have <sys/resource.h>.  Excellent.\n";
    return 1;
}
    
    # Solaris 1 weirdo header

    if (try_compile_and_link(<<EOM)) {
#include <sys/rusage.h>
static int foo;
EOM
    print "You have <sys/rusage.h>.  Are you an old Solaris by any chance?\n";
    return 2;
}

    return 0;
}

sub init {
    die <<EOM unless has_resource_h();
Your operating system does not seem to have <sys/resource.h> or <sys/rusage.h>.

(Run "perl Makefile.PL" with the environment variable VERBOSE set to 1 (one)
 too see in more detail what went wrong.)

There is no way $PACKAGE is going to work.

I am awfully sorry but I cannot go further.

Aborting configuration of $PACKAGE.
EOM

    $DEFINE = '';

    # Needed for NCR MP-RAS; should be harmless for others.
    if ($Config{'osname'} eq 'svr4') {
      $DEFINE = ' -D_XOPEN_SOURCE -D_XOPEN_SOURCE_EXTENDED=1';
    }

    # old Solarises had this oddity.
    $DEFINE .= ' -DI_SYS_RUSAGE' if (-r '/usr/include/sys/rusage.h');

    # ucb is poison for Solaris, and bsd may be on IRIX.
    if ($Config{'osname'} eq 'solaris') {
      $LIBS = [''];
    } elsif (($Config{'osname'} eq 'irix') &&
	     ($Config{'ccflags'} =~ m/-D_BSD_TIME/) &&
	     ($Config{'ccflags'} =~ m/-D_BSD_TYPES/)) {
      $LIBS = [''];
    } elsif ($Config{'osname'} eq 'svr4') {
      $LIBS = ['-lc89'];
    }
}

sub find_Rlim_t {
	my $tmp      = "rlim$$";
	my $not      = 0;

	print "Checking to see whether you have rlim_t...\n";
	my $has = try_compile_and_link(<<EOM);
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#ifdef I_SYS_TYPES
#include <sys/types.h>
#endif
#include <sys/resource.h>
static rlim_t foo = 0;
EOM
        if ($has) {
	    $DEFINE .= ' -DRlim_t=rlim_t';
	} else {
	    print "not ";
	    $not = 1;
	}

	print "You seem ";
	print "not " if $not;
	print "to have rlim_t defined (";
	print $not ? "this is okay" : "which is nice";
	print ").\n\n";
}

sub configure {
    find_Rlim_t;
    if ($^O eq 'solaris' &&
	$Config{ccflags} =~ /-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64/ &&
	$Config{use64bitall} ne 'define') {
	if (open(PROCFS, "/usr/include/sys/procfs.h")) {
	    my $noprocfs;
	    while (<PROCFS>) {
		if (/^#error.*Cannot use.*procfs.*large file.*/) {
		    $noprocfs = 1;
		    last;
		}
	    }
	    close(PROCFS);
	    if ($noprocfs) {
		$DEFINE .= ' -DSOLARIS_NO_PROCFS';
	    }
	}
    }
}

sub doMakefile {
    print  <<EOM if ($$LIBS[0] ne '');
Looking for libraries...
Note: it is ok if none of the libraries '@$LIBS' is found.

EOM
    WriteMakefile(
    'NAME'		=> $PACKAGE,
    'VERSION_FROM'	=> 'Resource.pm',
    'LIBS'		=> $LIBS,
    'DEFINE'		=> $DEFINE,
# do not even think about 'INC' => '-I/usr/ucbinclude', Solaris will avenge.
    'dist'      	=> { 'COMPRESS' => 'gzip' },
);
}

sub main {
    print <<EOM;

Configuring $PACKAGE...

EOM

    init;
    configure;
    doMakefile;
print  <<EOM;

Done configuring $PACKAGE.

Now you may issue 'make'.  Do not forget also 'make test'.
If some tests fail, please rerun the tests (see INSTALL).

EOM
}

&main;

# EOF


Resource.pm
#
# Copyright (c) 1995-2013 Jarkko Hietaniemi. All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
# Resource.pm
#

require 5.002;

package BSD::Resource;

use strict;
use vars qw(@ISA @EXPORT @EXPORT_OK $AUTOLOAD $VERSION);

$VERSION = '1.2906';

use Carp;
use AutoLoader;

require Exporter;
require DynaLoader;

@ISA = qw(Exporter DynaLoader);

@EXPORT = qw(
	     PRIO_MAX
	     PRIO_MIN
	     PRIO_PGRP
	     PRIO_PROCESS
	     PRIO_USER
	     RLIMIT_AIO_MEM
	     RLIMIT_AIO_OPS
	     RLIMIT_AS
	     RLIMIT_CORE
	     RLIMIT_CPU
	     RLIMIT_DATA
	     RLIMIT_FSIZE
	     RLIMIT_LOCKS
	     RLIMIT_MEMLOCK
	     RLIMIT_NOFILE
	     RLIMIT_NPROC
	     RLIMIT_OFILE
	     RLIMIT_OPEN_MAX
	     RLIMIT_RSS
	     RLIMIT_STACK
	     RLIMIT_TCACHE
	     RLIMIT_VMEM
	     RLIM_INFINITY
	     RLIM_NLIMITS
	     RLIM_SAVED_CUR
	     RLIM_SAVED_MAX
	     RUSAGE_BOTH
	     RUSAGE_CHILDREN
	     RUSAGE_SELF
	     RUSAGE_THREAD
	     get_rlimits
	     getpriority
	     getrlimit
	     getrusage	
	     setpriority
	     setrlimit
);

Exporter::export_tags();

@EXPORT_OK = qw(times);

# Grandfather old foo_h form to new :foo_h form
sub import {
    my $this = shift;
    my @list = map { m/^\w+_h$/ ? ":$_" : $_ } @_;
    local $Exporter::ExportLevel = 1;
    Exporter::import($this,@list);
}

bootstrap BSD::Resource;

my $EINVAL = constant("EINVAL", 0);
my $EAGAIN = constant("EAGAIN", 0);

sub AUTOLOAD {
    if ($AUTOLOAD =~ /::(_?[a-z])/) {
        $AutoLoader::AUTOLOAD = $AUTOLOAD;
        goto &AutoLoader::AUTOLOAD;
    }
    local $! = 0;
    my $constname = $AUTOLOAD;
    $constname =~ s/.*:://;
    return if $constname eq 'DESTROY';
    my $val = constant($constname, @_ ? $_[0] : 0);
    no strict 'refs';
    if ($! == 0) {
        *$AUTOLOAD = sub { $val };
    }
    elsif ($! == $EAGAIN) {     # Not really a constant, so always call.
        *$AUTOLOAD = sub { constant($constname, $_[0]) };
    }
    elsif ($! == $EINVAL) {
        croak "$constname is not a valid BSD::Resource macro";
    }
    else {
        croak "Your vendor has not defined BSD::Resource macro $constname, used";
    }
    use strict 'refs';

    goto &$AUTOLOAD;
}
use strict;

=pod

=head1 NAME

BSD::Resource - BSD process resource limit and priority functions

=head1 SYNOPSIS

	use BSD::Resource;

	#
	# the process resource consumption so far
	#

	($usertime, $systemtime,
	 $maxrss, $ixrss, $idrss, $isrss, $minflt, $majflt, $nswap,
	 $inblock, $oublock, $msgsnd, $msgrcv,
	 $nsignals, $nvcsw, $nivcsw) = getrusage($ru_who);

	$rusage = getrusage($ru_who);

	#
	# the process resource limits
	#

	($nowsoft, $nowhard) = getrlimit($resource);

	$rlimit = getrlimit($resource);

	$success = setrlimit($resource, $newsoft, $newhard);

	#
	# the process scheduling priority
	#

	$nowpriority = getpriority($pr_which, $pr_who);

	$success = setpriority($pr_which, $pr_who, $priority);

	# The following is not a BSD function.
	# It is a Perlish utility for the users of BSD::Resource.

	$rlimits = get_rlimits();

=head1 DESCRIPTION

=head2 getrusage

	($usertime, $systemtime,
	 $maxrss, $ixrss, $idrss, $isrss, $minflt, $majflt, $nswap,
	 $inblock, $oublock, $msgsnd, $msgrcv,
	 $nsignals, $nvcsw, $nivcsw) = getrusage($ru_who);

	$rusage = getrusage($ru_who);

	# $ru_who argument is optional; it defaults to RUSAGE_SELF

	$rusage = getrusage();

The $ru_who argument is either C<RUSAGE_SELF> (the current process) or
C<RUSAGE_CHILDREN> (all the child processes of the current process)
or it maybe left away in which case C<RUSAGE_SELF> is used.

The C<RUSAGE_CHILDREN> is the total sum of all the so far
I<terminated> (either successfully or unsuccessfully) child processes:
there is no way to find out information about child processes still
running.

On some systems (those supporting both getrusage() with the POSIX
threads) there can also be C<RUSAGE_THREAD>. The BSD::Resource supports
the C<RUSAGE_THREAD> if it is present but understands nothing more about
the POSIX threads themselves.  Similarly for C<RUSAGE_BOTH>: some systems
support retrieving the sums of the self and child resource consumptions
simultaneously.

In list context getrusage() returns the current resource usages as a
list. On failure it returns an empty list.

The elements of the list are, in order:
	index	name		meaning usually (quite system dependent)

	 0	utime		user time
	 1	stime		system time
    	 2	maxrss		maximum shared memory or current resident set
	 3	ixrss		integral shared memory
	 4	idrss		integral or current unshared data
	 5	isrss		integral or current unshared stack
	 6	minflt		page reclaims
	 7	majflt		page faults
    	 8	nswap		swaps
	 9	inblock		block input operations
	10	oublock		block output operations
	11	msgsnd		messages sent
	12	msgrcv		messaged received
	13	nsignals	signals received
	14	nvcsw		voluntary context switches
	15	nivcsw		involuntary context switches

In scalar context getrusage() returns the current resource usages as a
an object. The object can be queried via methods named exactly like
the middle column, I<name>, in the above table.

	$ru = getrusage();
	print $ru->stime, "\n";

	$total_context_switches = $ru->nvcsw + $ru->nivcsw;

For a detailed description about the values returned by getrusage()
please consult your usual C programming documentation about
getrusage() and also the header file C<E<lt>sys/resource.hE<gt>>.
(In B<Solaris>, this might be C<E<lt>sys/rusage.hE<gt>>).

See also L</"KNOWN ISSUES">.

=head2 getrlimit

	($nowsoft, $nowhard) = getrlimit($resource);

	$rlimit = getrlimit($resource);

The $resource argument can be one of

        $resource               usual meaning           usual unit

        RLIMIT_CPU              CPU time                seconds

        RLIMIT_FSIZE            file size               bytes

        RLIMIT_DATA             data size               bytes
        RLIMIT_STACK            stack size              bytes
        RLIMIT_CORE             coredump size           bytes
        RLIMIT_RSS              resident set size       bytes
        RLIMIT_MEMLOCK          memory locked data size bytes

        RLIMIT_NPROC            number of processes     1

        RLIMIT_NOFILE           number of open files    1
        RLIMIT_OFILE            number of open files    1
        RLIMIT_OPEN_MAX         number of open files    1

        RLIMIT_LOCKS            number of file locks    1

        RLIMIT_AS               (virtual) address space bytes
        RLIMIT_VMEM             virtual memory (space)  bytes

        RLIMIT_TCACHE           maximum number of       1
                                cached threads

        RLIMIT_AIO_MEM          maximum memory locked   bytes
                                for POSIX AIO
        RLIMIT_AIO_OPS          maximum number          1
                                for POSIX AIO ops

B<What limits are available depends on the operating system>.
See below for C<get_rlimits()> on how to find out which limits are
available, for the exact documentation consult the documentation of
your operating system.  The two groups (C<NOFILE>, C<OFILE>,
C<OPEN_MAX>) and (C<AS>, C<VMEM>) are aliases within themselves.

Two meta-resource-symbols might exist

	RLIM_NLIMITS
	RLIM_INFINITY

C<RLIM_NLIMITS> being the number of possible (but not necessarily fully
supported) resource limits, see also the get_rlimits() call below.
C<RLIM_INFINITY> is useful in setrlimit(), the C<RLIM_INFINITY> is
often represented as minus one (-1).

In list context C<getrlimit()> returns the current soft and hard
resource limits as a list.  On failure it returns an empty list.

Processes have soft and hard resource limits.  On crossing the soft
limit they receive a signal (for example the C<SIGXCPU> or C<SIGXFSZ>,
corresponding to the C<RLIMIT_CPU> and C<RLIMIT_FSIZE>, respectively).
The processes can trap and handle some of these signals, please see
L<perlipc/Signals>.  After the hard limit the processes will be
ruthlessly killed by the C<KILL> signal which cannot be caught.

B<NOTE>: the level of 'support' for a resource varies. Not all the systems

	a) even recognise all those limits
	b) really track the consumption of a resource
	c) care (send those signals) if a resource limit is exceeded

Again, please consult your usual C programming documentation.

One notable exception for the better: officially B<HP-UX> does not
support getrlimit() at all but for the time being, it does seem to.

In scalar context C<getrlimit()> returns the current soft limit.
On failure it returns C<undef>.

=head2 getpriority

	$nowpriority = getpriority($pr_which, $pr_who);

	# the default $pr_who is 0 (the current $pr_which)

	$nowpriority = getpriority($pr_which);

	# the default $pr_which is PRIO_PROCESS (the process priority)

	$nowpriority = getpriority();

getpriority() returns the current priority. B<NOTE>: getpriority()
can return zero or negative values completely legally. On failure
getpriority() returns C<undef> (and C<$!> is set as usual).

The priorities returned by getpriority() are in the (inclusive) range
C<PRIO_MIN>...C<PRIO_MAX>.  The $pr_which argument can be any of
PRIO_PROCESS (a process) C<PRIO_USER> (a user), or C<PRIO_PGRP> (a
process group). The $pr_who argument tells which process/user/process
group, 0 signifying the current one.

Usual values for C<PRIO_MIN>, C<PRIO_MAX>, are -20, 20.  A negative
value means better priority (more impolite process), a positive value
means worse priority (more polite process).

=head2 setrlimit

	$success = setrlimit($resource, $newsoft, $newhard);

setrlimit() returns true on success and C<undef> on failure.

B<NOTE>: A normal user process can only lower its resource limits.
Soft or hard limit C<RLIM_INFINITY> means as much as possible, the
real hard limits are normally buried inside the kernel and are B<very>
system-dependent.

B<NOTE>: Even the soft limit that is actually set might be lower than
what requested for various reasons.  One possibility is that the
actual limit on a resource might be controlled by some system variable
(e.g. in BSD systems the RLIMIT_NPROC can be capped by the system
variable C<maxprocperuid>, try C<sysctl -a kern.maxprocperuid>),
or in many environments core dumping has been disabled from normal
user processes.  Another possibility is that a limit is rounded down
to some alignment or granularity, for example the memory limits might
be rounded down to the closest 4 kilobyte boundary.  In other words,
do not expect to be able to setrlimit() a limit to a value and then be
able to read back the same value with getrlimit().

=head2 setpriority

	$success = setpriority($pr_which, $pr_who, $priority);

	# NOTE! If there are two arguments the second one is
	# the new $priority (not $pr_who) and the $pr_who is
	# defaulted to 0 (the current $pr_which)

	$success = setpriority($pr_which, $priority);

	# The $pr_who defaults to 0 (the current $pr_which) and
	# the $priority defaults to half of the PRIO_MAX, usually
	# that amounts to 10 (being a nice $pr_which).

	$success = setpriority($pr_which);

	# The $pr_which defaults to PRIO_PROCESS.

	$success = setpriority();

setpriority() is used to change the scheduling priority.  A positive
priority means a more polite process/process group/user; a negative
priority means a more impolite process/process group/user.
The priorities handled by setpriority() are [C<PRIO_MIN>,C<PRIO_MAX>].
A normal user process can only lower its priority (make it more positive).

B<NOTE>: A successful call returns C<1>, a failed one C<0>.

See also L</"KNOWN ISSUES">.

=head2 times

	use BSD::Resource qw(times);

	($user, $system, $child_user, $child_system) = times();

The BSD::Resource module offers a times() implementation that has
usually slightly better time granularity than the times() by Perl
core.  The time granularity of the latter is usually 1/60 seconds
while the former may achieve submilliseconds.

B<NOTE>: The current implementation uses two getrusage() system calls:
one with RUSAGE_SELF and one with RUSAGE_CHILDREN.  Therefore the
operation is not `atomic': the times for the children are recorded
a little bit later.

B<NOTE>: times() is not imported by default by BSD::Resource.
You need to tell that you want to use it.

B<NOTE: times() is not a "real BSD" function.  It is older UNIX.>

=head2 get_rlimits

	use BSD::Resource qw{get_rlimits};
	my $limits = get_rlimits();

B<NOTE: This is not a real BSD function. It is a convenience function
introduced by BSD::Resource.>

get_rlimits() returns a reference to hash which has the names of the
available resource limits as keys and their indices (those which
are needed as the first argument to getrlimit() and setrlimit())
as values. For example:

	use BSD::Resource qw{get_rlimits};
	my $limits = get_rlimits();
	for my $name (keys %$limits) {
	  my ($soft, $hard) = BSD::Resource::getrlimit($limits->{$name});
	  print "$name soft $soft hard $hard\n";
	}

Note that a limit of -1 means unlimited.

=head1 ERRORS

=over 4

=item *

	Your vendor has not defined BSD::Resource macro ...

The code tried to call getrlimit/setrlimit for a resource limit that
your operating system vendor/supplier does not support.  Portable code
should use get_rlimits() to check which resource limits are defined.

=back

=head1 EXAMPLES

	# the user and system times so far by the process itself

	($usertime, $systemtime) = getrusage();

	# ditto in OO way

	$ru = getrusage();

	$usertime   = $ru->utime;
	$systemtime = $ru->stime;

	# get the current priority level of this process

	$currprio = getpriority();

=head1 KNOWN ISSUES

In B<AIX> (at least version 3, maybe later also releases) if the BSD
compatibility library is not installed or not found by the BSD::Resource
installation procedure and when using the getpriority() or setpriority(),
the C<PRIO_MIN> is 0 (corresponding to -20) and C<PRIO_MAX> is 39
(corresponding to 19, the BSD priority 20 is unreachable).

In B<HP-UX> the getrusage() is not Officially Supported at all but for
the time being, it does seem to be.

In B<Mac OS X> a normal user cannot raise the C<RLIM_NPROC> over the
maxprocperuid limit (the default value is 266, try the command
C<sysctl -a kern.maxprocperuid>).

In B<NetBSD> C<RLIMIT_STACK> setrlimit() calls fail.

In B<Cygwin> C<RLIMIT_STACK> setrlimit calls fail.  Also, setrlimit()
C<RLIMIT_NOFILE/RLIMIT_OFILE/RLIMIT_OFILE> calls return success, but
then the subsequent getrlimit calls show that the limits didn't really
change.

Because not all UNIX kernels are BSD and also because of the sloppy
support of getrusage() by many vendors many of the getrusage() values
may not be correctly updated.  For example B<Solaris 1> claims in
C<E<lt>sys/rusage.hE<gt>> that the C<ixrss> and the C<isrss> fields
are always zero.  In B<SunOS 5.5 and 5.6> the getrusage() leaves most
of the fields zero and therefore getrusage() is not even used, instead
of that the B</proc> interface is used.  The mapping is not perfect:
the C<maxrss> field is really the B<current> resident size instead of the
maximum, the C<idrss> is really the B<current> heap size instead of the
integral data, and the C<isrss> is really the B<current> stack size
instead of the integral stack.  The ixrss has no sensible counterpart
at all so it stays zero.

=head1 COPYRIGHT AND LICENSE

Copyright 1995-2013 Jarkko Hietaniemi All Rights Reserved

This module free software; you can redistribute it and/or modify it
under the terms of the Artistic License 2.0 or GNU Lesser General
Public License 2.0.  For more details, see the full text of the
licenses at <http://www.perlfoundation.org/artistic_license_2_0>,
and <http://www.gnu.org/licenses/gpl-2.0.html>.

=head1 AUTHOR

Jarkko Hietaniemi, C<jhi@iki.fi>

=cut

1;
__END__

sub getrusage (;$) {
    my @rusage = _getrusage(@_);

    if (wantarray) {
	@rusage;
    } else {
	my $rusage = {};
	my $key;

	for $key (qw(utime stime maxrss ixrss idrss isrss minflt majflt nswap
		     inblock oublock msgsnd msgrcv nsignals nvcsw nivcsw)) {
	    $rusage->{$key} = shift(@rusage);
	}
	
	bless $rusage;
    }
}

sub _g {
    exists $_[0]->{$_[1]} ?
	$_[0]->{$_[1]} : die "BSD::Resource: no method '$_[1]',";
}

sub utime    { _g($_[0], 'utime'   ) }
sub stime    { _g($_[0], 'stime'   ) }
sub maxrss   { _g($_[0], 'maxrss'  ) }
sub ixrss    { _g($_[0], 'ixrss'   ) }
sub idrss    { _g($_[0], 'idrss'   ) }
sub isrss    { _g($_[0], 'isrss'   ) }
sub minflt   { _g($_[0], 'minflt'  ) }
sub majflt   { _g($_[0], 'majflt'  ) }
sub nswap    { _g($_[0], 'nswap'   ) }
sub inblock  { _g($_[0], 'inblock' ) }
sub oublock  { _g($_[0], 'oublock' ) }
sub msgsnd   { _g($_[0], 'msgsnd'  ) }
sub msgrcv   { _g($_[0], 'msgrcv'  ) }
sub nsignals { _g($_[0], 'nsignals') }
sub nvcsw    { _g($_[0], 'nvcsw'   ) }
sub nivcsw   { _g($_[0], 'nivcsw'  ) }

sub _find_rlimit ($$) {
    my ($func, $lim) = @_;
    if ($lim =~ /^RLIMIT_/) {
	my $rlimits = get_rlimits();
	if (exists $rlimits->{$lim}) {
	    $lim = $rlimits->{$lim};
	}
    }
    if ($lim =~ /^\d+$/) {
	# Looks fine.
    } else {
	croak "$func: Unknown limit '$lim'";
    }
    return $lim;
}

sub _find_prio ($$) {
    my ($func, $lim) = @_;
    if ($lim =~ /^PRIO_/) {
	my $prios = get_prios();
	if (exists $prios->{$lim}) {
	    $lim = $prios->{$lim};
	}
    }
    if ($lim =~ /^\d+$/) {
	# Looks fine.
    } else {
	croak "$func: Unknown limit '$lim'";
    }
    return $lim;
}

sub getrlimit ($) {
    my $lim = _find_rlimit('getrlimit', $_[0]);
    my @rlimit = _getrlimit($lim);

    if (wantarray) {
	return @rlimit;
    } else {
	return $rlimit[0];
    }
}

sub soft   { _g($_[0], 'soft') }
sub hard   { _g($_[0], 'hard') }

sub get_rlimits () {
    _get_rlimits();
}

sub get_prios () {
    _get_prios();
}

sub getpriority (;$$) {
    my ($which, $who) = @_;
    if (@_) {
	$which = _find_prio('getpriority', $which);
    }
    if (@_ == 2) {
	_getpriority($which, $who);
    } elsif (@_ == 1) {
	_getpriority($which);
    } else {
	_getpriority();
    }
}

sub setrlimit ($$$) {
    my ($lim, $soft, $hard) = @_;
    $lim = _find_rlimit('setrlimit', $lim);
    _setrlimit($lim, $soft, $hard);
}

sub setpriority (;$$$) {
    my ($which, $who, $prio) = @_;
    if (@_) {
	$which = _find_prio('setpriority', $which);
    }
    if (@_ == 3) {
	_setpriority($which, $who, $prio);
    } elsif (@_ == 2) {
	_setpriority($which, $who);
    } elsif (@_ == 1) {
	_setpriority($which);
    } else {
	_setpriority();
    }
}

sub times {
    use BSD::Resource qw(RUSAGE_SELF RUSAGE_CHILDREN);

    my ($u,  $s ) = _getrusage(RUSAGE_SELF);
    my ($cu, $cs) = _getrusage(RUSAGE_CHILDREN);

    return ($u, $s, $cu, $cs);
}

1;
